From a155fac91e534627c86399cc296f29426cf8436f Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 20 Oct 2022 14:47:52 +0300
Subject: [PATCH] [SW-108230] gaudi3: fix revoke bit computation

Allowed revoke bit values are 0, 1, 3, 7, and 15. In current code
when fuse bit is set to 0x10, ROM computes it to 0x0 which will
allow all keys to work. This patch adds a fix.

Change-Id: I32650e0032b3e2b0c7937d7ed2f3569135ba2690
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 2ebabf4..1e6f548 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -547,9 +547,6 @@
 	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_EOM_MASK)
 		arcboot_rom_global.config_fuse |= EFUSE_SYSCFG_PRE_PRODUCTION_BITS;
 
-	/* KEY index 4 must not be revoked */
-	arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_REVOKE_KEY_4;
-
 	/*
 	 * Validate key revokation bits. Expected values from fuse are
 	 * 0, 1, 3, 7, 15. If any other values are read from fuse override
@@ -560,8 +557,11 @@
 		arcboot_rom_global.config_fuse |=  EFUSE_SYSCFG_REVOKE_KEY_0_1;
 	else if ((key_revoke < 7) && (key_revoke > 3))
 		arcboot_rom_global.config_fuse |=  EFUSE_SYSCFG_REVOKE_KEY_0_1_2;
-	else if ((key_revoke < 15) && (key_revoke > 7))
+	else if ((key_revoke < 17) && (key_revoke > 7))
 		arcboot_rom_global.config_fuse |=  EFUSE_SYSCFG_REVOKE_KEY_0_1_2_3;
+
+	/* KEY index 4 must not be revoked */
+	arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_REVOKE_KEY_4;
 }
 
 __weak int arcboot_rom_fuse_single_die_config(void)
